### PR TITLE
Autoupdate

### PR DESCRIPTION
**Pull request by tiwiz**
This routine checks for new ClassyShark version every time the app starts. If a new JAR is released on GitHub, it will download it and it should override the present one. At the moment, the full process is silent and the user doesn't see a warning when the file has been downloaded.

### CODE REVIEW

File: ClassySharkWS/src/com/google/classyshark/Main.java [hunk 1]
```diff
@@ -18,6 +18,8 @@
 
 import com.google.classyshark.cli.CliMode;
 import com.google.classyshark.gui.GuiMode;
+import com.google.classyshark.updater.UpdateManager;
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/Main.java]: Seems the imports redundant, please remove

File: ClassySharkWS/src/com/google/classyshark/updater/UpdateManager.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import com.google.classyshark.updater.networking.AbstractReleaseCallback;
+import com.google.classyshark.updater.networking.MessageRunnable;
+import com.google.classyshark.updater.networking.NetworkManager;
+import com.google.classyshark.updater.utils.FileUtils;
+import retrofit2.Call;
+
+import javax.swing.*;
+import java.io.IOException;
+
+public class UpdateManager{
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/UpdateManager.java:35]: if this class is a Singleton, please do an enum class

File: ClassySharkWS/src/com/google/classyshark/updater/UpdateManager.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    public void checkVersionConsole() {
+        checkVersion(false);
+    }
+
+    public void checkVersionGui() {
+        checkVersion(true);
+    }
+
+    private void checkVersion(boolean gui) {
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/UpdateManager.java]: nit pick gui --> isGui

File: ClassySharkWS/src/com/google/classyshark/updater/models/Assets.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.classyshark.updater.models;
+
+class Assets {
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/models/Assets.java]: it is not clear what Assets is maybe call url XXXHolder and please make  browser_download_url camelCase

File: ClassySharkWS/src/com/google/classyshark/updater/models/Release.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+ * limitations under the License.
+ */
+
+package com.google.classyshark.updater.models;
+
+
+import com.google.classyshark.Version;
+import com.google.gson.annotations.SerializedName;
+import com.sun.istack.internal.Nullable;
+
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/models/Release.java:24]: Please add here a comment, about what does this class mean and how we do versioning.

File: ClassySharkWS/src/com/google/classyshark/updater/networking/GitHubApi.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package com.google.classyshark.updater.networking;
+
+
+import com.google.classyshark.updater.models.Release;
+import retrofit2.Call;
+import retrofit2.http.GET;
+
+public interface GitHubApi {
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/networking/GitHubApi.java:28]: please add a super short comment, what does this class do

File: ClassySharkWS/src/com/google/classyshark/updater/networking/NetworkManager.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * limitations under the License.
+ */
+
+package com.google.classyshark.updater.networking;
+
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+
+public class NetworkManager {
+
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/networking/NetworkManager.java:23]: please add private constructor

File: ClassySharkWS/src/com/google/classyshark/updater/utils/FileUtils.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        ReadableByteChannel rbc = Channels.newChannel(url.openStream());
+        File file = new File(NamingUtils.buildNameFrom(release));
+        file.getParentFile().mkdirs();
+        FileOutputStream fos = new FileOutputStream(file);
+        fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);
+        fos.close();
+        return file;
+    }
+
+    private void overwriteOld(File file) throws IOException {
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/utils/FileUtils.java:64]: can we remove overwriteOld as we don't do over-write old file

File: ClassySharkWS/src/com/google/classyshark/updater/utils/FileUtils.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+
+import static com.google.classyshark.updater.utils.NamingUtils.extractCurrentPath;
+import static javax.script.ScriptEngine.FILENAME;
+
+public class FileUtils {
+
+    public static File downloadFileFrom(Release release) throws IOException {
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/utils/FileUtils.java:47]: need to move that method under some other class

File: ClassySharkWS/src/com/google/classyshark/updater/networking/MessageRunnable.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.classyshark.updater.networking;
+
+import javax.swing.*;
+
+public class MessageRunnable implements Runnable {
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/networking/MessageRunnable.java:21]: Lets call the class DownloadCompleteRunnable

File: ClassySharkWS/src/com/google/classyshark/updater/networking/MessageRunnable.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    private final String title;
+    private final String changelog;
+    private final boolean gui;
```
Comment (borisf) [inline ClassySharkWS/src/com/google/classyshark/updater/networking/MessageRunnable.java:24]: nitpick isGui

**Comment by tiwiz** [comment]
The new update message looks like this

screen shot 2016-04-30 at 23 49 59

**Pull request merged**